orttraining/orttraining/training_ops/cuda/communication/nccl_service.cu

```diff
@@ -0,0 +1,197 @@
+    task = schedule_[time_].EqueueTask(NcclTask::Type::RECV, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&]{return task->is_finished;});
+  }
+};
+
+void NcclService::Initialize() {
+  // Here we assume GPU i is assigned to local process i.
+  // TODO: Create a general class to describe for computation topology and unify all similar uses.
+  // Hardwoards a process can own:
+  //   GPUs
+  //   CPUs
+  //   Other devices
+  int mpi_rank;
+  int mpi_size;
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+
+  // Set device this NCCL communicator runs on.
+  cudaSetDevice(mpi_rank);
+
+  // Create stream to run NCCL.
+  cudaStreamCreate(&stream_);
+
+  // Get NCCL unique ID at rank 0 and broadcast it to all others.
+  ncclUniqueId id;
+  if (mpi_rank == 0) ncclGetUniqueId(&id);
+  MPI_Bcast((void *)&id, sizeof(id), MPI_BYTE, 0, MPI_COMM_WORLD);
+  ncclCommInitRank(&comm_, mpi_size, id, mpi_rank);
+}
+
+void NcclService::Launch() {
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (!is_planned_) {
+      throw std::runtime_error("NCCL service must know its communication plan before launching.");
+    }
+    // The NCCL service object can only be launched once because it's a
+    // singlton class.
+    if (is_launched_) {
+      throw std::runtime_error("NCCL service cannot be repeatedly launched.");
+    }
+
+    // Set this flag so that others will not call this again.
+    is_launched_ = true;
+  }
+
+  Initialize();
+
+  while (is_launched_) {
+    // Enter critical region.
+    // The state of this class cannot be modified by other threads.
+    {
+      std::lock_guard<std::mutex> guard(mutex_);
+      // All tasks must be ready with a valid time.
+      if (time_ > schedule_.size() - 1 ||
+          !schedule_[time_].IsAllTasksEqueued() ||
+          schedule_[time_].IsAllTasksFinished()) {
+        continue;
+      }
+
+      // Start NCCL parallel communication.
+      ncclGroupStart();
+      for (auto& task : schedule_[time_].batch) {
+        if (!task.is_enqueued) {
+          throw std::runtime_error("Unscheduled task cannot be run.");
+        }
+
+        switch (task.type) {
+          case NcclTask::Type::SEND:
+            if (task.peers.size() != 1) {
+              throw std::invalid_argument("Send can only send data to one rank.");
+            }
+            ncclSend(task.ptr, task.size, ncclChar, task.peers.front(), comm_, stream_);
+            break;
+          case NcclTask::Type::RECV:
+            if (task.peers.size() != 1) {
+              throw std::invalid_argument("Recv can only send data to one rank.");
+            }
+            ncclRecv(task.ptr, task.size, ncclChar, task.peers.front(), comm_, stream_);
+            break;
+          default:
+            throw std::runtime_error("NCCL service currently only support ncclSend and ncclRecv.");
```

Comment:
The error handling style is different from other ORT code. Could use macros like ORT_ENFORCE.